olp-cpp-sdk-authentication/src/AuthenticationClient.cpp

```diff
@@ -235,6 +243,75 @@ IntrospectAppResult GetIntrospectAppResult(const rapidjson::Document& doc) {
   return result;
 }
 
+std::vector<ActionResult> GetDiagnostics(
+    rapidjson::Document& doc,
+    const std::function<DecisionType(std::string)>& get_permission) {
+  std::vector<ActionResult> results;
+    const auto& array = doc[Constants::DIAGNOSTICS].GetArray();
+    for (auto& element : array) {
+      ActionResult action;
+      if (element.HasMember(Constants::DECISION)) {
+        action.SetDecision(
+            get_permission(element[Constants::DECISION].GetString()));
+        // get permissions if avialible
+        if (element.HasMember(Constants::PERMISSIONS) &&
+            element[Constants::PERMISSIONS].IsArray()) {
+          std::vector<ActionResult::Permissions> permissions;
+          const auto& permissions_array =
+              element[Constants::PERMISSIONS].GetArray();
+          for (auto& permission_element : permissions_array) {
+            ActionResult::Permissions permission;
+            if (permission_element.HasMember(Constants::ACTION)) {
+              permission.first =
+                  permission_element[Constants::ACTION].GetString();
+            }
+            if (permission_element.HasMember(Constants::DECISION)) {
+              permission.second = get_permission(
+                  permission_element[Constants::DECISION].GetString());
+            }
+            permissions.push_back(std::move(permission));
+          }
+
+          action.SetPermissions(std::move(permissions));
+        }
+      }
+      results.push_back(std::move(action));
+    }
+    return results;
+}
+
+AuthorizeResult GetAuthorizeResult(rapidjson::Document& doc) {
+  AuthorizeResult result;
+
+  auto get_permission = [](std::string str) -> DecisionType {
```

Comment:
Instead of using lamdba i would use a anonymous function to avoid passing the lambda to GetDiagnostics().